Accommodate sample datasets that are ZIP archives

niksirbi opened the pull request:
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

So far, all of our loading functions assume that the tracking output (predicted poses or bboxes) is stored in a single file.
However, some software—like TRex—store data across multiple files (in TRex's case, one file per individual). See #599 for details on this.

The solution, as prototyped by @albiangela in this [colab notebook](https://colab.research.google.com/drive/1vvFPMWrHlLsnPOul8LdsmYqbmy-Y6sWp?usp=sharing), is to have our TRex loading function accept a folder path (instead of a file path), and load all relevant files from that folder.

To help integrate this feature into `movement`, it would be helpful to add some TRex data to our [test data repository](https://gin.g-node.org/neuroinformatics/movement-test-data). I've already added a dataset called `TRex_five-locusts.zip`, which is a zipped folder containing 5 .npz files.

The problem, is that our sample data infrastructure assumes that each dataset is always a single file and isn't set up to handle zipped archives.

**What does this PR do?**

Modifies `sample_data.py` to also handle zipped archives.

Specifically: if the requested sample dataset name ends with `.zip`, the archive is unpacked after download (using the `pooch.Unzip()` processor), and `sample_data.fetch_dataset_paths()` returns a path to the unzipped folder that contains the constituent files. This new functionality is implemented as a `_fetch_and_unzip_folder` utility.

You can test this behaviour by running:
```python
>>>  from movement import sample_data

>>>  paths = sample_data.fetch_dataset_paths("TRex_five-locusts.zip")
>>>  print(paths["poses"])
/Users/<USER>/.movement/data/poses/TRex_five-locusts.zip.unzip/TRex_five-locusts   # This is a folder
```

Calling `sample_data.fetch_dataset("TRex_five-locusts.zip")` will raise a `NotImplementedError`, as we haven't yet implemented the `load_poses.from_trex_file()` function. Once that function is implemented, `sample_data.fetch_dataset("TRex_five-locusts.zip")` will work the same as for the other datasets, and we can remove the `NotImplementedError`.


## References

Facilitates  #599.
Please read that issue to get the full context.

## How has this PR been tested?

Tests in `test_sample_data.py` have been modified to account for the new functionality.

## Is this a breaking change?

No. All existing sample datasets should work the same as before.

## Does this PR require an update to the documentation?

I have added a paragraph to the contributing guide.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

lochhh commented on movement/sample_data.py:
```diff
@@ -304,3 +318,41 @@ def fetch_dataset(
     return ds
+
+
+def _fetch_and_unzip_folder(data_type: str, filename: str) -> Path:
+    """Download & unpack a zipped folder containing multiple files.
+
+    Parameters
+    ----------
+    data_type : str
+        Type of data to fetch, e.g. "poses", "bboxes"
+    filename : str
+        Name of the zipped folder to fetch, e.g. "TRex_five-locusts.zip"
+
+    Returns
+    -------
+    Path
+        Path to the folder containing the unpacked files,
+        e.g. "/path/to/TRex_five-locusts.zip.unzip/TREX_five-locusts/"
+
+    """
+    raw_paths = SAMPLE_DATA.fetch(
+        f"{data_type}/{filename}",
+        processor=pooch.Unzip(),
+        progressbar=True,
+    )  # list of str
+    paths = [Path(p) for p in raw_paths]  # convert to Path objects
+
+    # filter out unwanted files and directories
+    UNWANTED_FILES = {".DS_Store", "Thumbs.db", "desktop.ini"}
+    UNWANTED_DIRS = {"__MACOSX"}
+
+    filtered = [
+        p
+        for p in paths
+        if p.name not in UNWANTED_FILES
+        and not any(part in UNWANTED_DIRS for part in p.parts)
+    ]
+    # all remaining files should live in the same folder
+    return filtered[0].parent
```
I think we could do away with "folder" in the function name. We could also go a step further to remove the unwanted files and directories, while keeping the .zip file (since this is what pooch uses to determine if the file needs updating). Rather than recursively searching and deleting in the `unzip` dir, the strategy here is just to `shutil.move` (need to `import shutil`) the `filtered` files to a new destination directory with the same name as the original `.zip` file and deleting the `unzip` dir:
```suggestion
def _fetch_and_unzip(data_type: str, file_name: str | Path) -> Path:
    """Download and extract a zipped archive and return the folder path.

    Parameters
    ----------
    data_type : str
        Type of data to fetch, e.g. "poses", "bboxes".
    file_name : str
        Name of the .zip file to fetch, e.g. "TRex_five-locusts.zip"

    Returns
    -------
    Path
        Path to the folder containing the unarchived file(s),
        e.g. "/path/to/TRex_five-locusts/"

    """
    file_path = Path(data_type) / file_name
    raw_paths = SAMPLE_DATA.fetch(
        file_path.as_posix(),
        processor=pooch.Unzip(),
        progressbar=True,
    )
    paths = [Path(p) for p in raw_paths]  # convert to Path objects

    # Filter data files and move them to the destination directory
    UNWANTED_FILES = {".DS_Store", "Thumbs.db", "desktop.ini"}
    UNWANTED_DIRS = {"__MACOSX"}
    extract_dir_name = file_path.with_suffix("").name
    extract_dir = SAMPLE_DATA.path / data_type / f"{file_name}.unzip"
    dest_dir = SAMPLE_DATA.path / data_type / extract_dir_name
    dest_dir.mkdir(parents=True, exist_ok=True)
    for path in paths:
        if path.name in UNWANTED_FILES or any(
            part in UNWANTED_DIRS for part in path.parts
        ):
            continue
        try:
            rel_path = path.relative_to(extract_dir / extract_dir_name)
            dest_path = dest_dir / rel_path
            dest_path.parent.mkdir(parents=True, exist_ok=True)
            shutil.move(str(path), str(dest_path))
        except Exception as e:
            logger.warning(
                f"Failed to move file {path} to {dest_path}. "
                f"Using the original path instead: {e}"
            )
            return path.parent

    # Clean up the .unzip folder
    try:
        shutil.rmtree(extract_dir)
    except Exception as e:
        logger.warning(f"Failed to remove unzip dir {extract_dir}: {e}")
    return dest_dir
```

niksirbi replied:
I'm aware of that, it will become a problem if we even host bigger datasets as .zip. I will see if there's away to rewrite this that avoids the aforementioned "strange" situation.

lochhh replied:
Tbf the "strange" situation could happen both ways - you could also succeed in copying one file but not another, for instance if the permissions are incorrect.

niksirbi replied:
This is what I have now, with corresponding tests.
This version is robust against any of the strange situations, but adds the copy overhead.
It works, but sometimes I think whether we are too defensive, might have been fine to just move all the valid files and let it just fail if it errors (which is very unlikely to happen imo).

```python
def _fetch_and_unzip(data_type: str, file_name: str | Path) -> Path:
    """Download and extract a zipped archive and return the folder path.

    Parameters
    ----------
    data_type : str
        Type of data to fetch, e.g. "poses", "bboxes".
    file_name : str
        Name of the .zip file to fetch, e.g. "TRex_five-locusts.zip"

    Returns
    -------
    Path
        Path to the folder containing the unarchived file(s),
        e.g. "/path/to/TRex_five-locusts/"

    """
    file_path = Path(data_type) / file_name
    raw_paths = SAMPLE_DATA.fetch(
        file_path.as_posix(),
        processor=pooch.Unzip(),
        progressbar=True,
    )
    paths = [Path(p) for p in raw_paths]  # convert to Path objects

    # Filter data files
    UNWANTED_FILES = {".DS_Store", "Thumbs.db", "desktop.ini"}
    UNWANTED_DIRS = {"__MACOSX"}
    valid_paths = [
        path
        for path in paths
        if path.name not in UNWANTED_FILES
        and not any(part in UNWANTED_DIRS for part in path.parts)
    ]

    # Copy files to a new destination directory
    extract_dir_name = file_path.with_suffix("").name
    extract_dir = SAMPLE_DATA.path / data_type / f"{file_name}.unzip"
    dest_dir = SAMPLE_DATA.path / data_type / extract_dir_name
    dest_dir.mkdir(parents=True, exist_ok=True)

    try:
        for path in valid_paths:
            rel_path = path.relative_to(extract_dir / extract_dir_name)
            dest_path = dest_dir / rel_path
            dest_path.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy2(path.as_posix(), dest_path.as_posix())
    except Exception as e:
        logger.warning(
            f"Failed to copy files from {extract_dir} to {dest_dir}. "
            f"Using the original path instead: {e}"
        )
        shutil.rmtree(dest_dir)  # remove the dest_dir
        return valid_paths[0].parent  # return the original path

    # Clean up the .unzip folder
    try:
        shutil.rmtree(extract_dir)
    except Exception as e:
        logger.warning(f"Failed to remove unzip dir {extract_dir}: {e}")
    return dest_dir
```

lochhh commented on tests/test_unit/test_sample_data.py:
```diff
@@ -1,42 +1,49 @@
+        "fps": None,
+        "frame_file": None,
+        "video_file": None,
+    },
```
We should add TRex here as well
```suggestion
    },
    "TRex_five-locusts.zip": {
        "fps": 5,
        "frame_file": "locusts-noqr_n-5_date-20250117_frame-bg.png",
        "video_file": None,
    },
```

lochhh commented on tests/test_unit/test_sample_data.py:
```diff
@@ -1,42 +1,49 @@
+}
 
 
 @pytest.fixture(scope="module")
 def valid_sample_datasets():
```
We can parametrise the fixture
```suggestion
@pytest.fixture(params=list(SAMPLE_DATASETS.items()))
def sample_dataset(request):
    """Return the name of a sample dataset and its metadata."""
    sample_name, sample_metadata = request.param
    return sample_name, sample_metadata
```

lochhh commented on tests/test_unit/test_sample_data.py:
```diff
@@ -1,42 +1,49 @@
-            "video_file": None,
-        },
-    }
+    """Return the predefined sample datasets mapping."""
+    return SAMPLE_DATASETS.copy()
```
```suggestion
```

lochhh commented on tests/test_unit/test_sample_data.py:
```diff
@@ -127,28 +134,79 @@ def test_list_datasets(valid_sample_datasets):
     assert all(file in list_datasets() for file in valid_sample_datasets)
```
This needs to be updated if we decide to parametrise the `sample_dataset` fixture
```suggestion
def test_list_datasets():
    assert isinstance(list_datasets(), list)
    assert all(file in list_datasets() for file in SAMPLE_DATASETS)
```

lochhh commented on tests/test_unit/test_sample_data.py:
```diff
@@ -127,28 +134,79 @@ def test_list_datasets(valid_sample_datasets):
     assert all(file in list_datasets() for file in valid_sample_datasets)
 
 
+# Parametrized fetch_dataset test
+@pytest.mark.parametrize(
+    "sample_name, sample_metadata",
+    list(SAMPLE_DATASETS.items()),
+    ids=list(SAMPLE_DATASETS.keys()),
+)
```
This can be replaced by passing in `sample_dataset` an arg
```suggestion
```

lochhh commented on tests/test_unit/test_sample_data.py:
```diff
@@ -127,28 +134,79 @@ def test_list_datasets(valid_sample_datasets):
-    # Test with an invalid file
-    with pytest.raises(ValueError):
-        fetch_dataset("nonexistent_file")
+def test_fetch_dataset(sample_name, sample_metadata, with_video):
+    """Test fetch_dataset for each sample dataset with/without video."""
```
```suggestion
def test_fetch_dataset(sample_dataset, with_video):
    """Test fetch_dataset for each sample dataset with/without video."""
    sample_name, sample_metadata = sample_dataset
    # Skip TRex datasets as they are not supported yet
    if sample_name.startswith("TRex"):
        pytest.xfail("TRex datasets are not supported yet.")
```

lochhh commented on tests/test_unit/test_sample_data.py:
```diff
@@ -127,28 +134,79 @@ def test_list_datasets(valid_sample_datasets):
+        fetch_dataset(sample_name)
+
+
+@pytest.mark.parametrize(
+    "sample_name",
+    [
+        list(SAMPLE_DATASETS.keys())[0],
+        "TRex_five-locusts.zip",
+    ],
+    ids=[
+        "poses_in_single_file",
+        "poses_in_zipped_folder",
+    ],
+)
+def test_fetch_dataset_paths(sample_name):
+    """Test that the returned pose paths points to correct location.
+
+    If the pose files are in a zipped folder, the path should point to the
+    unzipped folder, otherwise it should point to the file itself.
+    """
+    paths = fetch_dataset_paths(sample_name)
+    poses_path = Path(paths["poses"])
+
+    if sample_name.endswith(".zip"):
+        # If the sample is a zip file,
+        # the path should point to the unzipped folder
+        assert poses_path.is_dir()
+        assert poses_path.name == sample_name.replace(".zip", "")
+        assert len(list(poses_path.iterdir())) > 1
+    else:
+        # If the sample is a single file,
+        # the path should point to the file itself
+        assert poses_path.is_file()
+        assert poses_path.name == sample_name
```
This will cover all 5 test samples we have, which I suppose isn't a bad idea
```suggestion
def test_fetch_dataset_paths(sample_dataset):
    """Test that the returned pose paths points to correct location.

    If the pose files are in a zipped folder, the path should point to the
    unzipped folder, otherwise it should point to the file itself.
    """
    sample_name, _ = sample_dataset
    paths = fetch_dataset_paths(sample_name)
    data_path = Path(paths.get("poses", paths.get("bboxes")))

    if sample_name.endswith(".zip"):
        # If the sample is a zip file,
        # the path should point to the unzipped folder
        assert data_path.is_dir()
        assert data_path.name == sample_name.replace(".zip", "")
        assert len(list(data_path.iterdir())) > 1
    else:
        # If the sample is a single file,
        # the path should point to the file itself
        assert data_path.is_file()
        assert data_path.name == sample_name
```

lochhh commented on CONTRIBUTING.md:
```diff
@@ -437,65 +441,44 @@ The most important parts of this module are:
 2. The `list_datasets()` function, which returns a list of the available poses and bounding boxes datasets (file names of the data files).
 3. The `fetch_dataset_paths()` function, which returns a dictionary containing local paths to the files associated with a particular sample dataset: `poses` or `bboxes`, `frame`, `video`. If the relevant files are not already cached locally, they will be downloaded.
-4. The `fetch_dataset()` function, which downloads the files associated with a given sample dataset (same as `fetch_dataset_paths()`) and additionally loads the pose or bounding box data into movement, returning an `xarray.Dataset` object. If available, the local paths to the associated video and frame files are stored as dataset attributes, with names `video_path` and `frame_path`, respectively.
+4. The `fetch_dataset()` function, which downloads the files associated with a given sample dataset (same as `fetch_dataset_paths()`) and additionally loads the pose or bounding box data into `movement`, returning an `xarray.Dataset` object. If available, the local paths to the associated video and frame files are stored as dataset attributes, with names `video_path` and `frame_path`, respectivel ...
```
do we want to link the module and these functions to the api ref?

lochhh replied:
btw there's inconsistency throughout the file with monospacing `movement`.

niksirbi replied:
Yeah, I noticed that, but I only corrected those in the sample data section. I'm thinking of enforcing the consistency in a separate PR...

lochhh replied:
Sounds good to me! Can be resolved by opening an issue.

niksirbi replied:
Already started a PR, which will need to be rebased after this one is merged, see https://github.com/neuroinformatics-unit/movement/pull/626

lochhh commented on CONTRIBUTING.md:
```diff
@@ -437,65 +441,44 @@ The most important parts of this module are:
+4. Clone the `movement` data repository to your local machine using `gin get neuroinformatics/movement-test-data`, then run `gin download --content` to download all the files.
+5. Add your new files to the appropriate folders (`poses`, `bboxes`, `videos`, and/or `frames`) following the existing file naming conventions.
+6. Add metadata for your new files to `metadata.yaml` using the example entry below as a template. You can leave all `sha256sum` values as `null` for now.
+7. Update file hashes in `metadata.yaml` by running `python update_hashes.py` from the root of the [movement data repository](gin:neuroinformatics/movement-test-data). This script computes SHA256 hashes for all data files and updates the corresponding `sha256sum` values in the metadata file. Make sure you're in a Python environment with `movement` installed.
```
maybe link "a Python environment with `movement` installed" to the env setup?

niksirbi replied:
done

lochhh commented on CONTRIBUTING.md:
```diff
@@ -437,65 +441,44 @@ The most important parts of this module are:
+6. Add metadata for your new files to `metadata.yaml` using the example entry below as a template. You can leave all `sha256sum` values as `null` for now.
+7. Update file hashes in `metadata.yaml` by running `python update_hashes.py` from the root of the [movement data repository](gin:neuroinformatics/movement-test-data). This script computes SHA256 hashes for all data files and updates the corresponding `sha256sum` values in the metadata file. Make sure you're in a Python environment with `movement` installed.
+8. Commit your changes using `gin commit -m <message> <filename>` for specific files or `gin commit -m <message> .` for all changes.
+9. Upload your committed changes to the GIN repository with `gin upload`. Use `gin download` to pull the latest changes or `gin sync` to synchronize changes bidirectionally.
```
```suggestion
9. Upload your committed changes to the GIN repository with `gin upload`. Use `gin download` to pull the latest changes or `gin sync` to synchronise changes bidirectionally.
```

lochhh reviewed (approved):
Thanks @niksirbi ! The PR looks good. I do have a few suggestions here and there for you to consider:
- deleting unwanted files after unzipping
- parametrising the `sample_dataset` fixture 
- cross-referencing functions in the contributing guide

niksirbi commented:
Thanks for the thorough review @lochhh.

- I've basically adopted all of your suggestions for the tests, which were excellent.
- For `_fetch_and_unzip()`, I ended up going with [this version](https://github.com/neuroinformatics-unit/movement/pull/619#discussion_r2175249881) in the end. I've also added a test for the various failure modes of this utility (copy fails or rmtree fails). Let me know if you'd prefer a different solution.
- I added crossrefs to the contributing guide, as suggested
- I started a separate PR #626

lochhh commented on tests/test_unit/test_sample_data.py:
```diff
@@ -122,33 +141,140 @@ def test_fetch_metadata(tmp_path, caplog, download_fails, local_exists):
+
+    It also tests that unwanted files are not copied to the new dest folder.
+    """
+    from movement import sample_data
```
we already import from sample_data at the top, might as well import SAMPLE_DATA and _fetch_and_unzip there

lochhh commented on tests/test_unit/test_sample_data.py:
```diff
@@ -122,33 +141,140 @@ def test_fetch_metadata(tmp_path, caplog, download_fails, local_exists):
+    monkeypatch.setattr(
+        sample_data.SAMPLE_DATA,
+        "fetch",
+        lambda *a, **k: [file_to_copy.as_posix(), unwanted_file.as_posix()],
```
```suggestion
        lambda *a, **k: [
            str(extract_dir / file_name) for file_name in file_names
        ],
```

niksirbi replied:
You are right!

lochhh commented on tests/test_unit/test_sample_data.py:
```diff
@@ -122,33 +141,140 @@ def test_fetch_metadata(tmp_path, caplog, download_fails, local_exists):
+        result_names = [p.name for p in result.iterdir()]
+        if failing_func == "copy2":
+            # original folder is kept, incl. unwanted file
+            assert result == file_to_copy.parent
```
```suggestion
            assert result == extract_dir
```

niksirbi commented:
All new suggestions implemented [in this commit](https://github.com/neuroinformatics-unit/movement/pull/619/commits/9cddc69d91bb2951b7e76c49ddf2fe2604f00e94) 🎉
